libraries/standard/mqtt/src/mqtt_lightweight.c

```diff
@@ -192,6 +252,513 @@ static int32_t recvExact( MQTTTransportRecvFunc_t recvFunc,
 
 /*-----------------------------------------------------------*/
 
+static size_t _getRemainingLength( MQTTTransportRecvFunc_t recvFunc,
+                                   MQTTNetworkContext_t networkContext )
+{
+    size_t remainingLength = 0, multiplier = 1, bytesDecoded = 0, expectedSize = 0;
+    uint8_t encodedByte = 0;
+    int32_t bytesReceived = 0;
+
+    /* This algorithm is copied from the MQTT v3.1.1 spec. */
+    do
+    {
+        if( multiplier > 2097152U ) /* 128 ^ 3 */
+        {
+            remainingLength = MQTT_REMAINING_LENGTH_INVALID;
+        }
+        else
+        {
+            bytesReceived = recvFunc( networkContext, &encodedByte, 1U );
+            if( bytesReceived == 1 )
+            {
+                remainingLength += ( ( size_t ) encodedByte & 0x7FU ) * multiplier;
+                multiplier *= 128U;
+                bytesDecoded++;
+            }
+            else
+            {
+                remainingLength = MQTT_REMAINING_LENGTH_INVALID;
+            }
+        }
+        if( remainingLength == MQTT_REMAINING_LENGTH_INVALID )
+        {
+            break;
+        }
+    } while ( ( encodedByte & 0x80U ) != 0U );
+
+    /* Check that the decoded remaining length conforms to the MQTT specification. */
+    if( remainingLength != MQTT_REMAINING_LENGTH_INVALID )
+    {
+        expectedSize = _remainingLengthEncodedSize( remainingLength );
+        if( bytesDecoded != expectedSize )
+        {
+            remainingLength = MQTT_REMAINING_LENGTH_INVALID;
+        }
+    }
+
+    return remainingLength;
+}
+
+/*-----------------------------------------------------------*/
+
+static bool _incomingPacketValid( uint8_t packetType )
+{
+    bool status = false;
+
+    /* Check packet type. Mask out lower bits to ignore flags. */
+    switch( packetType & 0xF0U )
+    {
+        /* Valid incoming packet types. */
+        case MQTT_PACKET_TYPE_CONNACK:
+        case MQTT_PACKET_TYPE_PUBLISH:
+        case MQTT_PACKET_TYPE_PUBACK:
+        case MQTT_PACKET_TYPE_PUBREC:
+        case MQTT_PACKET_TYPE_PUBCOMP:
+        case MQTT_PACKET_TYPE_SUBACK:
+        case MQTT_PACKET_TYPE_UNSUBACK:
+        case MQTT_PACKET_TYPE_PINGRESP:
+            status = true;
+            break;
+
+        /* Any other packet type is invalid. */
+        default:
+            break;
+    }
+
+    /* The second bit of a PUBREL must be set, so the above switch won't handle it. */
+    if( packetType == MQTT_PACKET_TYPE_PUBREL )
```

Comment:
I would prefer to see this moved into the switch. You can add it like this:
case MQTT_PACKET_TYPE_PUBREL & 0xF0 :
    if (packetType & 0x02) {
       status = true;
   }

I think it is easier to understand this way?

Reply:
Changed